sdks/python/apache_beam/io/gcp/bigquery_tools.py

```diff
@@ -1114,29 +1112,16 @@ def insert_rows(
     # BigQuery will do a best-effort if unique IDs are provided. This situation
     # can happen during retries on failures.
     # TODO(silviuc): Must add support to writing TableRow's instead of dicts.
-    final_rows = []
-    for i, row in enumerate(rows):
-      json_row = self._convert_to_json_row(row)
-      insert_id = str(self.unique_row_id) if not insert_ids else insert_ids[i]
-      final_rows.append(
-          bigquery.TableDataInsertAllRequest.RowsValueListEntry(
-              insertId=insert_id, json=json_row))
+    insert_ids = [
+        str(self.unique_row_id) if not insert_ids else insert_ids[i] for i,
+        _ in enumerate(rows)
+    ]
+    rows = [json.loads(json.dumps(r, default=default_encoder)) for r in rows]
```

Comment:
any reason to dump and load? sanity check?

Reply:
we need to pass a dictionary where its values are already json-encoded. What this line does is "encode-everything-into-json" and then "decode JSON into a dictionary with JSON-encoded fields"

Reply:
Okay. I thought `lambda x: loads(dumps(x))` is idempotent but looks like it's not the case here.